src/cargo/ops/registry.rs

```diff
@@ -459,24 +457,8 @@ fn registry(
     force_update: bool,
     validate_token: bool,
 ) -> CargoResult<(Registry, RegistryConfig, SourceId)> {
-    if index.is_some() && registry.is_some() {
-        // Otherwise we would silently ignore one or the other.
-        bail!("both `--index` and `--registry` should not be set at the same time");
-    }
-    // Parse all configuration options
+    let (sid, sid_no_replacement) = get_source_id(config, index, registry)?;
     let reg_cfg = registry_configuration(config, registry)?;
-    let opt_index = registry
-        .map(|r| config.get_registry_index(r))
-        .transpose()?
-        .map(|u| u.to_string());
-    let sid = get_source_id(config, opt_index.as_deref().or(index), registry)?;
-    if !sid.is_remote_registry() {
-        bail!(
-            "{} does not support API commands.\n\
```

Comment:
What happened to this code? Do we still get a good error message if you call an API on a registry that does not support API commands?

Reply:
Yes, we still get the same error message if the registry doesn't support API commands. See a few lines below
```
        cfg.and_then(|cfg| cfg.api)
            .ok_or_else(|| format_err!("{} does not support API commands", sid))?
```

This wasn't really checking if the source supported API commands, it was checking if it was a non-remote registry. The replacement error message for this case is now more specific (and provides a better suggestion):
```
crates-io is replaced with non-remote-registry source {replacement_sid};
include `--registry crates-io` to publish to crates.io
```